### ISSUE TITLE
meta-data to identify macros that introduce new symbols

### ISSUE THREAD

**Issue by hlship** [cursive-ide/cursive#147]
In reference to letfn, etc., would it be possible to introduce a meta-data value that would identify which parameter, in a macro, was to be introduced as a symbol or binding form?

```
(defmacro let-like [^:binding-form bindings & body] .... )
```

When Cursive is analyzing this code:

```
  (let-like [foo bar
              baz biff] (x foo baz))
```

It could use the macro definition to identify that foo and baz are symbols that can be referenced int he body of the form.

Possibly, we'd need an additional meta-data value for the body itself.

It's a thorny problem.

**Comment by atroche** [comment]
Any update on what the workaround's going to be?

**Comment by jugimaster** [comment]
Can you introduce some sort of "stop-gap solution"? People have had this problem ever since Cursive came out.

**Comment by trimtab613** [comment]
Does emacs do this? If so, Cursive should absolutely be able to. If not, then let's give up early.

**Comment by cursiveclojure** [comment]
@trimtab613 Emacs doesn't do anything like this, no, but that's no reason to give up.

**Comment by aboy021** [comment]
Is the solution to this related to the talk "[Improving Clojure's Error Messages with Grammars - Colin Fleming - YouTube](https://www.youtube.com/watch?v=kt4haSH2xcs)"?

If so, perhaps there could mechanism for registering macro specific grammars with cursive?

You could have a couple of base cases that support things like the aforementioned "behaves like defn" and "behaves like let".

Ideally this would be stored in some sort of standard, stand alone, format so that if library contributors wanted to add the functionality it would be simple and non-invasive to do so. Potentially other tools could make use of the grammars then too.

**Comment by guv** [comment]
Is there any (planned) work on a solution, yet? A quickfix where the Cursive user can specify the macro semantics would be a good start. Currently, there is a lot of noise from "cannot be resolved" messages in projects using custom "def*" macros and other very useful features of Cursive don't work with functions defined via custom macros (documentation view, auto-completion, ...).

**Comment by trimtab613** [comment]
My company is pretty much split between Emacs and Cursive, so inline indicators probably wouldn't go over well in the Emacs crowd. I put my vote behind a global or per-project configuration file that can be generated bit by bit by a command in the context menu. Something like "Manually Define Symbol", which would display a dialog asking for details, then writes the info to that config file.

**Comment by guv** [comment]
Just as a clarification: I vote for a macro semantic definition outside the actual code as well.
Ideally, a per library configuration file (maybe distributed with the library) or a global configuration (per Cursice install) as a quickfix.

**Comment by jzwolak** [comment]
fbellomi, if I'm including a library in my project and expanding the macro while editing formats my hard drive then that just moves the problem from when I run my code to when I edit my code. I trust the libraries I import in to my project, so there is no need to sandbox them while macro expanding to discover defined symbols. I'm going to run the code anyway, might as well run it while editing and get the benefit of symbol resolution.

The halting problem is a real one, but most IDEs already have functions to address things like this during analysis and other background tasks. NetBeans for instance automatically detects slowness in running tasks and reports this to the user. Such an approach could be used to give this symbol resolution feature x amount of time to complete and if it doesn't complete then a warning is issued to the user and the analysis is either cancelled or the user is given the option to cancel/continue. The x amount of time can be per macro expansion.

I really like the idea of solving this problem without adding an API and meta data.

With meta data the burden is passed to humans... let's not burden ourselves to adding meta data for eternity (every new macro that defines symbols will need someone to type the metadata) if we can write a solution once and the computer can work on it for us.

Yes, it is a challenge to write an automated solution in Cursive. That's exactly the kind of challenge I believe the Clojure community is up for and exactly what sets Clojure apart: good solutions to tough problems instead of lazy solutions that are almost good.

**Comment by cursive-ide** [comment]
@guv Yes, I'm planning to try to implement this for the next major release, hopefully within a month.

@jzwolak It's not a lazy solution, it's a pragmatic one. Even expanding macros isn't sufficient in some cases, for example Storm creates a bunch of symbols from the public members of a class. This means that in order to do that, the analysis process requires the whole project to be compiled and the classpath set correctly - it's far from trivial and would lead to terrible performance. Creating inverted indexes is hard from runtime inspection. The API is not just for symbol resolution but will be required to provide information for other aspects of Cursive's functionality (type inference, method implementation creation etc) which cannot be determined by macro expansion alone.

**Comment by gnl** [comment]
@cursive-ide Is the implementation you're planning going to work in ClojureScript as well?

**Comment by cursive-ide** [comment]
Yes, definitely.

**Comment by cursive-ide** [comment]
@aboy021 Sorry, I missed this message earlier. Yes, this is related to my talk, and the current internal solutions use grammars as shown there.

**Comment by guv** [comment]
@cursive-ide That is great news! :)

**Comment by jzwolak** [comment]
@cursive-ide, thanks for your thoughts on this. I hope I wasn't sounding too critical by using the words "lazy solution". I do also see the need for pragmatic solutions.

The whole project only needs to be compiled once in the example with Storm that you mention. IntelliJ takes about that amount of time building indices either at startup or when something significant changes in the project. Once this initial compilation is performed and all symbols are resolved only the changed file and files depending on it need recompiling. For most projects and most cases this will only be a handful of files. If the developer is working on Storm itself and modifying those files with macros that scan classes for methods to make Clojure functions from then presumably in this case the developer would have the option of turning off this feature for this project... or use this API we're talking about. I was reading somewhere about the recommendation for restraint in using macros... perhaps it was in the "Joy of Clojure". There are developers of libraries and developers of applications, my intuition is that macros will and should appear less in applications and are more useful in libraries. I understand the performance issues and I get this is a challenging feature to automatically scan for these dependencies... my current opinion is leaning towards a hybrid after reading your message and thinking about it. The API would be very useful in the Storm case, but the automatic analysis is not impossible and would be very useful in the majority of cases where macro expansion and recompiling a couple files is involved.

**Comment by cursive-ide** [comment]
I've made the first steps towards macro support in the latest Cursive release (1.3.0-eap1, out now). Details are [here](https://cursive-ide.com/userguide/macros.html), basically you can select a problematic macro form and say "this works like `let`" or "this works like `def`". It's pretty limited so far, but just with those simple cases I hope to solve perhaps 70-80% of the pain of this problem. Feedback welcome on whether this achieves that. Full support is on the way.

**Comment by sfnelson** [comment]
Thanks for this Colin, it's a big improvement. One major problem with the current incarnation is that these settings are stored in a user-specific location, rather than project specific. This means these settings cannot be easily shared amongst team members.

It would be extremely helpful if these settings could be configured per project, and stored in the project's `.idea` directory, e.g. `codeStyleSettings.xml`.

**Comment by guv** [comment]
How is the general solution for macros progressing?

**Comment by guv** [comment]
Since it has been a while: Has the general solution remained a goal or has it been abandoned?

**Comment by cursive-ide** [comment]
The general solution remains a goal, definitely. I've made some steps towards this such as the "Resolve as..." functionality and the stubs generation. The full solution is a lot of work, but I'm working towards it bit by bit.

**Comment by a613** [comment]
Still awaiting this. We use some libraries that generate symbols with macros (e.g. soda-ash), and we've got warnings about undefined symbols all over our code because of it. The emacs crowd is winning because their IDE actually recognizes that the symbols are used, and can even jump to the macro they're defined in.

**Comment by cursive-ide** [comment]
@a613 Libraries like soda-ash are exactly what the stub generation was designed for. I'll add that to the list, and try to bump up making the list configurable so you can add that sort of lib yourself.

**Comment by sfnelson** [comment]
While you're looking at macros, please take pity on https://github.com/cursive-ide/cursive/issues/953

**Comment by MrOerni** [comment]
Anything new on the topic of custom resolving?
Maybe just a simple solution like a check against a regex for every symbol?
Or the option to disable the resolving for a particular macro?

----
Additionaly:
I have something like this:

```clojure
(defaction my-action
  :contexts [a b]
  :parameters [c d]
  (do-something-with a b c d)
```
This may be a complicated matter as the symbols themselves are not known for the specific macro. 
I wonder If something like this could be done with `spec`.

**Comment by guv** [comment]
This is stalled for quite some time.
Might there be a more pragmatic approach compared to your current envisioned solution for custom macro support?

**Comment by p-himik** [comment]
Another, and IMO the best, approach is to have a section similar to `Settings | Editor | Language Injections` that would allow customizing the way Cursive resolves unknown symbols.
Ideally, such a mechanism would:
1. Allow explicitly specifying whether a particular resolution is project-specific or not (i.e. global)
2. Save its settings into a file that could be easily tracked with git (so, along with similar files in `.idea`)
3. Allow import/export so that people could share the settings for their library

**Comment by aneilbaboo** [comment]
Given that this has been a very long running issue, could we get a fix that simply allows turning OFF symbol resolution for a specific macro? Perhaps this is already possible through the "resolve as" feature, but it's not obvious. It would be good if that were a top-level choice. E.g., In this menu, I'd like to be able to choose "disable symbol resolution for foo.bar/mymacro":
![image](https://user-images.githubusercontent.com/2592723/92329008-8259e700-f019-11ea-9f78-1639e829b73d.png)

**Comment by tekacs** [comment]
@aneilbaboo this was just resolved in #2417!
